authority/tls.go

```diff
@@ -180,34 +189,48 @@ func (a *Authority) Renew(oldCert *x509.Certificate) ([]*x509.Certificate, error
 		PolicyIdentifiers:           oldCert.PolicyIdentifiers,
 	}
 
-	// Copy all extensions except for Authority Key Identifier. This one might
-	// be different if we rotate the intermediate certificate and it will cause
-	// a TLS bad certificate error.
+	// Copy all extensions except:
+	//	1. Authority Key Identifier - This one might be different if we rotate the intermediate certificate
+	//					and it will cause a TLS bad certificate error.
+	//	2. Subject Key Identifier - This should be calculated for the public key passed to this function.
 	for _, ext := range oldCert.Extensions {
-		if !ext.Id.Equal(oidAuthorityKeyIdentifier) {
+		if (!ext.Id.Equal(oidAuthorityKeyIdentifier)) && (!ext.Id.Equal(oidSubjectKeyIdentifier)) {
 			newCert.ExtraExtensions = append(newCert.ExtraExtensions, ext)
 		}
+		if ext.Id.Equal(oidSubjectKeyIdentifier) {
```

Comment:
This (https://github.com/smallstep/cli/blob/master/crypto/x509util/profile.go#L305-L312) snippet of code will always set a SubjectKeyID for new profiles if one is not explicitly set in the template. The SubjectKeyID can be overwritten by an ExtraExtension at the time when the template is signed, so I think all we need to do is make sure we don't append the oidSubjectKeyIdentifier to the cert extra extensions. Let's let @maraino weigh in here before making any changes, but I think that would be easier.

Reply:
I checked this. Yes, It is needless to append the SubjectKeyIdentifier in tls.go.

Reply:
This is tricky because the x509 RFC recommends but does not enforce the use of that method for subject key identifiers. So at least right now we should be backward compatible with the previous implementation and keep the Subject Key Identifier as it was.

Reply:
After thinking a little bit I think we need two do different things here for renew and rekey:
1. For `Renew`, we want to keep the certificate intact, so we should not change the Subject Key Identifier
2. For `Rekey`, we should change the subject key identifier because is a way to detect all the certificates with the same key. I believe `x509util.NewLeafProfileWithTemplate` will set the SubjectKeyId if it's not set, so, for `Rekey` only, we should skip this extension and set `newCert.SubjectKeyId` to nil to force `x509util.NewLeafProfileWithTemplate` to create a new one.

A simple way to detect a Renew or Rekey, can be to pass the pk as nil in Renew and use the existing one.